scipy/spatial/transform/rotation.py

```diff
@@ -0,0 +1,874 @@
+            dcm_transformed[~safe2, 1, 0] + dcm_transformed[~safe2, 0, 1],
+            dcm_transformed[~safe2, 0, 0] - dcm_transformed[~safe2, 1, 1]
+        )
+
+    # Step 7
+    if seq[0] == seq[2]:
+        # lambda = 0, so we can only ensure angle2 -> [0, pi]
+        adjust_mask = np.logical_or(angles[:, 1] < 0, angles[:, 1] > np.pi)
+    else:
+        # lambda = + or - pi/2, so we can ensure angle2 -> [-pi/2, pi/2]
+        adjust_mask = np.logical_or(angles[:, 1] < -np.pi / 2,
+                                    angles[:, 1] > np.pi / 2)
+
+    # Dont adjust gimbal locked angle sequences
+    adjust_mask = np.logical_and(adjust_mask, safe_mask)
+
+    angles[adjust_mask, 0] += np.pi
+    angles[adjust_mask, 1] = 2 * offset - angles[adjust_mask, 1]
+    angles[adjust_mask, 2] -= np.pi
+
+    angles[angles < -np.pi] += 2 * np.pi
+    angles[angles > np.pi] -= 2 * np.pi
+
+    # Step 8
+    if not np.all(safe_mask):
+        warnings.warn("Gimbal lock detected. Setting third angle to zero since"
+                      " it is not possible to uniquely determine all angles.")
+
+    # Reverse role of extrinsic and intrinsic rotations, but let third angle be
+    # zero for gimbal locked cases
+    if extrinsic:
+        angles = angles[:, ::-1]
+    return angles
+
+
+def _make_elementary_quat(axis, angles):
+    quat = np.zeros((angles.shape[0], 4))
+
+    quat[:, 3] = np.cos(angles / 2)
+    quat[:, _AXIS_TO_IND[axis]] = np.sin(angles / 2)
+    return quat
+
+
+def _compose_quat(p, q):
+    product = np.empty((max(p.shape[0], q.shape[0]), 4))
+    product[:, 3] = p[:, 3] * q[:, 3] - np.sum(p[:, :3] * q[:, :3], axis=1)
+    product[:, :3] = (p[:, None, 3] * q[:, :3] + q[:, None, 3] * p[:, :3] +
+                      np.cross(p[:, :3], q[:, :3]))
+    return product
+
+
+def _elementary_quat_compose(seq, angles, intrinsic=False):
+    result = _make_elementary_quat(seq[0], angles[:, 0])
+
+    for idx, axis in enumerate(seq[1:], start=1):
+        if intrinsic:
+            result = _compose_quat(
+                result,
+                _make_elementary_quat(axis, angles[:, idx]))
+        else:
+            result = _compose_quat(
+                _make_elementary_quat(axis, angles[:, idx]),
+                result)
+    return result
+
+
+class Rotation(object):
+    """Rotation in 3 dimensions.
+
+    This class provides an interface to initialize from and represent rotations
+    with:
+
+        - Quaternions
+        - Direction Cosine Matrices
+        - Rotation Vectors
+        - Euler angles
+
+    The following operations on rotations are supported:
+
+        - Application on vectors
+        - Rotation Composition
+        - Rotation Inversion
+        - Rotation Indexing
+
+    Indexing within a rotation is supported since multiple rotation transforms
+    can be stored within a single `Rotation` instance.
+
+    Initialization using the `from_...` classmethods such as `from_quat` is
```

Comment:
I think we should say ``To create `Rotation` objects use `from_...` classmethods, `__init__` is not supposed to be used directly.``